Stratis.Bitcoin/BlockPulling/QualityScore.cs

```diff
@@ -104,28 +93,27 @@ public void AddSample(IBlockPullerBehavior peer, long blockDownloadTimeMs, int b
             lock (this.lockObject)
             {
                 // If we reached the maximum number of samples, we need to remove oldest sample.
-                if (this.samplesCount == this.samples.Length)
+                if (this.samples.Count == this.samples.Capacity)
                 {
-                    PeerSample oldSample = this.samples[this.samplesIndex];
+                    PeerSample oldSample = this.samples[this.samples.Index];
                     this.samplesSum -= oldSample.timePerKb;
                     this.peerReferenceCounter[oldSample.peer]--;
 
                     if (this.peerReferenceCounter[oldSample.peer] == 0)
                         this.peerReferenceCounter.Remove(oldSample.peer);
                 }
-                else this.samplesCount++;
 
                 // Add new sample to the mix.
-                this.samples[this.samplesIndex].timePerKb = timePerKb;
-                this.samples[this.samplesIndex].peer = peer;
-                this.samplesIndex = (this.samplesIndex + 1) % this.samples.Length;
+                int index = this.samples.AddNoSet();
```

Comment:
That was the whole purpose of using it - to avoid allocation. When it was inline, it was super simple, but @dangershony did not like the inline implementation. So I separated it only to find out that it is hard to avoid the allocation with the "standard" separated implementation (I've basically ended up with simpler version of Queue), so to gain back the full power of why I used it in the first place, I had to use AddNoSet.